### PR TITLE
LSTM/GRU/RNN Sequences : support for seq_lengths input

### PR DESCRIPTION
**Pull request by itikhono**
- Sequences to TI transformations
- Adjust reference implementations of GRU/LSTM/RNN Seqs
- fix ngraph python API
- IR Reader unit tests
- enable ONNX importer unit tests
- fix TensorIterator ref impl and tests

### CODE REVIEW

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/gru_sequence.cpp [hunk 1]
```diff
@@ -30,10 +34,11 @@ namespace {
 
     INSTANTIATE_TEST_CASE_P(smoke_GRUSequenceCommonZeroClip, GRUSequenceTest,
                             ::testing::Combine(
+                                    ::testing::ValuesIn(mode),
                                     ::testing::ValuesIn(seq_lengths_zero_clip),
                                     ::testing::ValuesIn(batch),
                                     ::testing::ValuesIn(hidden_size),
-                                    ::testing::ValuesIn(input_size),
+                                    // ::testing::ValuesIn(input_size), // hardcoded to 10 due to Combine supports up to 10 args
```
Comment (mikhail-treskin) [inline inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/gru_sequence.cpp:41]: In general, you can use nested `Combine`s to workaround such gtest limitation

Comment (itikhono) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/gru_sequence.cpp:41]: Could you provide some usage examples in code?

**Review by mikhail-treskin — CHANGES_REQUESTED** [review]

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/gru_sequence.cpp (hunk 1 above)
Comment (mikhail-treskin) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/gru_sequence.cpp:41]: https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/convolution.cpp

File: inference-engine/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/reverse_sequence.cpp [hunk 2]
```diff
@@ -41,7 +41,7 @@ INSTANTIATE_TEST_CASE_P(Basic_smoke, ReverseSequenceLayerTest,
                             ::testing::ValuesIn(reversSeqLengthsVecShapes),
                             ::testing::ValuesIn(secondaryInputTypes),
                             ::testing::ValuesIn(netPrecisions),
-                            ::testing::Values(CommonTestUtils::DEVICE_GPU)),
+                            ::testing::Values(CommonTestUtils::DEVICE_CPU)),
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/reverse_sequence.cpp]: fixed

**Review by mbencer — APPROVED** [review]

File: ngraph/python/src/ngraph/opset3/ops.py [hunk 3]
```diff
@@ -432,8 +432,8 @@ def rnn_cell(
     @param W:                       The weight tensor with shape: [hidden_size, input_size].
     @param R:                       The recurrence weight tensor with shape: [hidden_size,
                                     hidden_size].
-    @param B:                       The bias tensor for input gate with shape: [2*hidden_size].
-    @param hidden_size:             The number of hidden units for recurrent cell.
+    :param B:                       The sum of biases (weight and recurrence) with shape: [hidden_size].
+    :param hidden_size:             The number of hidden units for recurrent cell.
```
Comment (itikhono) [inline ngraph/python/src/ngraph/opset3/ops.py:436]: missprint

**Review by mitruska — APPROVED** [review]

File: ngraph/python/src/ngraph/opset3/ops.py (hunk 3 above)
Comment (itikhono) [reply ngraph/python/src/ngraph/opset3/ops.py:436]: @mbencer @postrational which option is correct "@" or ":"?

File: ngraph/test/onnx/onnx_import_rnn.in.cpp [hunk 4]
```diff
@@ -1668,7 +1668,7 @@ NGRAPH_TEST_F(${BACKEND_NAME}, RNNSequenceOp, onnx_model_rnn_fwd_mixed_seq_len_c
                                              -0.18203181f,
                                              0.9996245f,
                                          });
-    test_case.run(DEFAULT_FLOAT_TOLERANCE_BITS + 3);
+    test_case.run(DEFAULT_FLOAT_TOLERANCE_BITS + 4);
```
Comment (ilyachur) [inline ngraph/test/onnx/onnx_import_rnn.in.cpp:1671]: Why did you change the number?

File: inference-engine/tests/ngraph_functions/src/utils/ngraph_helpers.cpp [hunk 5]
```diff
@@ -255,6 +256,17 @@ std::vector<std::uint8_t> convertPrecision(std::vector<std::uint8_t> &buffer, co
     return convertedData;
 }
 
+bool is_tensor_iterator_exist(const std::shared_ptr<ngraph::Function> & func) {
+    const auto& ops = func->get_ops();
+    for (const auto& node : ops) {
+        const auto& ti = std::dynamic_pointer_cast<ngraph::opset5::TensorIterator>(node);
+        if (ti) {
```
Comment (ilyachur) [inline inference-engine/tests/ngraph_functions/src/utils/ngraph_helpers.cpp:263]:
```suggestion
        if (std::dynamic_pointer_cast<ngraph::opset5::TensorIterator>(node)) {
```

File: inference-engine/tests/ngraph_functions/src/rnn_cell.cpp [hunk 6]
```diff
@@ -27,9 +28,20 @@ std::shared_ptr<ngraph::Node> makeRNN(const OutputVector& in,
...
+            || mode == ngraph::helpers::SequenceTestsMode::CONVERT_TO_TI_MAX_SEQ_LEN_CONST) {
+            std::vector<float> lengths(in[0].get_shape()[0], in[0].get_shape()[1]);
+            seq_lengths = ngraph::builder::makeConstant(element::i64, constants[3], lengths, false);
+        } else if (mode == ngraph::helpers::SequenceTestsMode::CONVERT_TO_TI_RAND_SEQ_LEN_CONST) {
+            for (size_t  i = 0; i <= in[0].get_shape().at(0); ++i) {
+                std::vector<float> lengths;
+                seq_lengths = ngraph::builder::makeConstant(element::i64, constants[3], lengths, true, in[0].get_shape()[1], 0);
+            }
+        } else {
+            seq_lengths = in.at(2).get_node_shared_ptr();
```
Comment (ilyachur) [inline inference-engine/tests/ngraph_functions/src/rnn_cell.cpp:42]: `at` throws an exception if element doesn't exist. Are we sure that `in` vector has 3 elements.

File: inference-engine/tests/ngraph_functions/src/lstm_cell.cpp [hunk 7]
```diff
@@ -27,9 +28,20 @@ std::shared_ptr<ngraph::Node> makeLSTM(const std::vector<ngraph::Output<Node>>&
...
+            || mode == ngraph::helpers::SequenceTestsMode::CONVERT_TO_TI_MAX_SEQ_LEN_CONST) {
+            std::vector<float> lengths(in[0].get_shape()[0], in[0].get_shape()[1]);
+            seq_lengths = ngraph::builder::makeConstant(element::i64, constants[3], lengths, false);
+        } else if (mode == ngraph::helpers::SequenceTestsMode::CONVERT_TO_TI_RAND_SEQ_LEN_CONST) {
+            for (size_t  i = 0; i <= in[0].get_shape().at(0); ++i) {
+                std::vector<float> lengths;
+                seq_lengths = ngraph::builder::makeConstant(element::i64, constants[3], lengths, true, in[0].get_shape()[1], 0);
+            }
+        } else {
+            seq_lengths = in.at(3).get_node_shared_ptr();
```
Comment (ilyachur) [inline inference-engine/tests/ngraph_functions/src/lstm_cell.cpp:42]: `at` throws an exception if element doesn't exist. Are we sure that `in` vector has 4 elements.

File: ngraph/test/onnx/onnx_import_rnn.in.cpp (hunk 4 above)
Comment (itikhono) [reply ngraph/test/onnx/onnx_import_rnn.in.cpp:1671]: accuracy mismatch for 1 of 60 values. Other tests pass. This seems acceptable.

File: inference-engine/tests/ngraph_functions/src/rnn_cell.cpp (hunk 6 above)
Comment (itikhono) [reply inference-engine/tests/ngraph_functions/src/rnn_cell.cpp:42]: else branch only for modes == CONVERT_TO_TI_RAND_SEQ_LEN_PARAM, CONVERT_TO_TI_MAX_SEQ_LEN_PARAM . So it means Seq_len input should be as a Parameter here. But maybe better write an exception message instead of .at exception

File: inference-engine/tests/ngraph_functions/src/utils/ngraph_helpers.cpp (hunk 5 above)
Comment (itikhono) [reply inference-engine/tests/ngraph_functions/src/utils/ngraph_helpers.cpp:263]: will fix locally

File: inference-engine/tests/ngraph_functions/src/lstm_cell.cpp (hunk 7 above)
Comment (itikhono) [reply inference-engine/tests/ngraph_functions/src/lstm_cell.cpp:42]: else branch is only for modes == CONVERT_TO_TI_RAND_SEQ_LEN_PARAM, CONVERT_TO_TI_MAX_SEQ_LEN_PARAM . So it means Seq_len input should be as a Parameter here. But maybe it's better to write an exception message instead of .at exception

File: ngraph/test/onnx/onnx_import_rnn.in.cpp (hunk 4 above)
Comment (itikhono) [reply ngraph/test/onnx/onnx_import_rnn.in.cpp:1671]: it's a new test, it was disabled in manifest. Not a regression

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp [hunk 8]
```diff
@@ -0,0 +1,583 @@
...
+#include <ngraph/opsets/opset5.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/op/util/activation_functions.hpp>
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::ConvertRNNSequenceToTensorIterator, "ConvertRNNSequenceToTensorIterator", 0);
+NGRAPH_RTTI_DEFINITION(ngraph::pass::ConvertGRUSequenceToTensorIterator, "ConvertGRUSequenceToTensorIterator", 0);
+NGRAPH_RTTI_DEFINITION(ngraph::pass::ConvertLSTMSequenceToTensorIterator, "ConvertLSTMSequenceToTensorIterator", 0);
+
+ngraph::Output<ngraph::Node> get_current_iter(ngraph::ParameterVector& body_params,
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp:19]: Can we add all internal methods to unnamed namespace?

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp [hunk 9]
```diff
@@ -0,0 +1,583 @@
...
+             *   We need to set rnn_name.0 to RevSequence to store result name.
+             *   TI [rnn_name] -- (DATA_NODE: rnn_name.0) --> RevSequence [rnn_name.0] -- (DATA_NODE: rnn_name.0) -> Result1
+             *                 -- (data_node: rnn_name.1) --> Result2
+             *
+             *
+             *   After transformation (with identity, no collisions):
+             *   TI has other_name, but it doesn't affect result names due TI is not connected to Results directly.
+             *   TI [other_name] -- (data_node: other_name.0) --> RevSequence [rnn_name.0] -- (data_node: rnn_name.0) -> Result1
+             *                   -- (data_node: other_name.1) --> Identity(rnn_name.1) -- (data_node: rnn_name.1) -> Result2
+             */
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp:225]: Very complex logic! Thank you for the comments. Could you highlight this in the epic for outputs naming 37965?

**Review by ilyachur — APPROVED** [review]

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp [hunk 10]
```diff
@@ -0,0 +1,583 @@
...
+        Output<Node> h_node_to_result = unsqueeze_dum_dir;
+        if (enable_mask) {
+            auto current_iter = get_current_iter(body_params, body_results, seq_body_param);
+            h_node_to_result = get_masked_value(tensor_iterator, body_params, body_results, current_iter,
+                                                unsqueeze_dum_dir, seq_body_param);
+        }
+
+        auto H_res = std::make_shared<opset5::Result>(h_node_to_result);
+        auto unsqueeze_seq_len = std::make_shared<opset5::Unsqueeze>(h_node_to_result, axis_1);
+        auto concat_res = std::make_shared<opset5::Result>(unsqueeze_seq_len);
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp:157]: concat_res it's Result for concatenation, we should add an additional dimension to perform concat operation (via TI internal logic) along this additional axis.

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp [hunk 11]
```diff
@@ -0,0 +1,583 @@
...
+        NodeVector new_nodes;
+        const auto &X = sequence->input_value(0); // split
+        const auto &H_t = sequence->input_value(1); // merged (init value + back edge)
+        const auto &seq_lengths = sequence->input_value(2); // invariant
+        const auto &W = sequence->input_value(3); // const in the body
+        const auto &R = sequence->input_value(4); // const in the body
+        const auto &B = sequence->input_value(5); // const in the body
+        bool is_reverse = sequence->get_direction() == ngraph::op::RecurrentSequenceDirection::REVERSE;
+
+        if (X.get_partial_shape().is_dynamic() || H_t.get_partial_shape().is_dynamic()
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp:107]: >> BTW I realized that seq_lengths doesn't required to have static shape.
not sure, I will check

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp [hunk 12]
```diff
@@ -0,0 +1,583 @@
...
+
+ngraph::NodeVector squeeze_nodes(const ngraph::OutputVector& nodes_to_squeeze, const ngraph::OutputVector & axes) {
+    ngraph::NodeVector squeezed_nodes(nodes_to_squeeze.size());
+    for (size_t i = 0; i < nodes_to_squeeze.size(); ++i) {
+        squeezed_nodes[i] = std::make_shared<ngraph::opset5::Squeeze>(nodes_to_squeeze[i], axes[i]);
+    }
+    return squeezed_nodes;
+}
+
+bool should_enable_mask(const ngraph::Output<ngraph::Node>& seq_lengths, size_t max_seq_len) {
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp:73]: looks better, thanks

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp [hunk 13]
```diff
@@ -0,0 +1,583 @@
...
+        }
+
+        NodeVector new_nodes;
+        const auto &X = sequence->input_value(0); // split
+        const auto &H_t = sequence->input_value(1); // merged (init value + back edge)
+        const auto &seq_lengths = sequence->input_value(2); // invariant
+        const auto &W = sequence->input_value(3); // const in the body
+        const auto &R = sequence->input_value(4); // const in the body
+        const auto &B = sequence->input_value(5); // const in the body
+        bool is_reverse = sequence->get_direction() == ngraph::op::RecurrentSequenceDirection::REVERSE;
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp:105]: I will fix it

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp [hunk 14]
```diff
@@ -0,0 +1,583 @@
...
+            || seq_lengths.get_partial_shape().is_dynamic()) {
+            return false;
+        }
+
+        auto tensor_iterator = std::make_shared<opset5::TensorIterator>();
+        auto max_seq_len = X.get_shape().at(1);
+        bool enable_mask = should_enable_mask(seq_lengths, max_seq_len);
+
+        std::shared_ptr<Node> reverse_seq_before;
+        if (is_reverse && enable_mask) {
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp:117]: actually, it will be, checked

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp (hunk 8 above)
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp:19]: Sure

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp (hunk 9 above)
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp:225]: I have updated the description (Known issues list) in 37965, but I didn't create separate tickets.

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp:157]: Spec of LSTM/RNN/GRU Sequences:
* **1**: `Y`“ 3D tensor of type *T1* `[batch_size, num_directions, seq_len, hidden_size]`, concatenation of all the intermediate output values of the hidden.

* **2**: `Ho` - 3D tensor of type *T1* `[batch_size, num_directions, hidden_size]`, the last output value of hidden state.

for Y we add num_direction and seq_len single dimensions
for Ho only num_direction

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp [hunk 15]
```diff
@@ -0,0 +1,583 @@
...
+    const auto& data_shape = data.get_shape();
+    const auto& data_shape_size = ngraph::shape_size(data_shape);
+
+    // body parameters
+    auto aggregated_Y_h_body_param = std::make_shared<ngraph::opset5::Parameter>(data_type, data_shape);
+
+    body_params.push_back(aggregated_Y_h_body_param);
+
+    // Create mask node deciding whether or not to mask batch data.
+    ngraph::Output<ngraph::Node> batch_seq_length = ngraph::builder::opset1::legacy_broadcast_for_binary_operation(
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp:50]: opset5 ops are aliases to opset1, for all ops in this `legacy_broadcast_for_binary_operation` function.
I don't want to duplicate the code of this function and don't want to upgrade the function to opset5 due to there are other places of use of this function.

**This is a general problem in our code: we cannot create helper functions with a hard-coded version of the opset. It is a more complicated issue than using this function. I will open a question in our morning meeting.**

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/gru_sequence.cpp (hunk 1 above)
Comment (itikhono) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/gru_sequence.cpp:41]: So I also need to change GRU Cell tests, I will do it for both GRU Sequence/Cell in the next PR (ticket 43220)

File: ngraph/python/src/ngraph/opset3/ops.py (hunk 3 above)
Comment (itikhono) [reply ngraph/python/src/ngraph/opset3/ops.py]: fixed

File: inference-engine/tests/ngraph_functions/src/lstm_cell.cpp (hunk 7 above)
Comment (itikhono) [reply inference-engine/tests/ngraph_functions/src/lstm_cell.cpp]: fixed

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp [hunk 16]
```diff
@@ -0,0 +1,583 @@
...
+            data, seq_lengths, 0);
+    auto mask_value = std::make_shared<ngraph::opset5::Constant>(data.get_element_type(),
+                             data.get_shape(), std::vector<float>(data_shape_size, 0.f));
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp]: fixed

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp]: fixed

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp]: Seq_lenghts should be static

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp:110]: fixed

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp (hunk 8 above)
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp]: done

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_sequences_to_tensor_iterator.cpp]: And we also allow MO to generate ops from different opsets

**Review by mikhail-treskin — APPROVED** [review]

**Review by GlebKazantaev — APPROVED** [review]

**Comment by ilyachur** [comment]
@itikhono please resolve conflicts.

**Comment by dkurt** [comment]
Hi!  This PR breaks LSTM accuracy. See details in 42954.

**Comment by itikhono** [comment]
the accuracy issue above is not related to this PR, it is observed in previous commits
